rts/Game/UI/KeyBindings.cpp

```diff
@@ -757,14 +746,15 @@ bool CKeyBindings::AddKeySymbol(const std::string& keysym, const std::string& co
 }
 
 
-bool CKeyBindings::RemoveCommandFromList(ActionList& al, const std::string& command)
+bool CKeyBindings::RemoveActionFromList(ActionList& al, const Action& action, const Action::Comparison& comparison)
 {
 	bool success = false;
 
 	auto it = al.begin();
 
 	while (it != al.end()) {
-		if (it->command == command) {
+		const auto a = it.base();
+		if (comparison(action, *a)) {
```

Comment:
@sprunk Is this dereference costly? If so is there a way to make `comparison` maintain `const Action&, const Action& -> boolean` signature (compatible with `std::` functions) while maintaining performance?

Reply:
Not that remove from list is a critical perf code path, but I want to keep it as tidy as possible and learn so I know for future efforts on critical perf code paths.

Reply:
Dereference should be cheap.

I'm not sure what `base()` does here, what would happen if it was just `if (comparison(action, *it)) {`?

Also this should be able to be rewritten more idiomatically. Something like
```cpp
#include <algorithm>

bool CKeyBindings::RemoveActionFromList(ActionList& al, const Action& action, const Action::Comparison& comparison)
{
  // requires C++20
  const auto num_removed = std::erase_if(al, [](const Action &a) { return comparison(action, a); });
  return num_removed > 0;
  
  // alternatively, pre-20
  const auto rm_iter = std::remove_if(al.begin(), al.end(), [](const Action &a) { return comparison(action, a); });
  const bool any_removed = (rm_iter != al.end());
  al.erase(rm_iter, al.end());
  return any_removed;
}
```

Reply:
Ah yeah, I maintained most of the original authors code on the initial effort. Sounds good to me, thanks!

What would you call it instead?

Reply:
> I'm not sure what base() does here

It comes from the iterator definition, just a reference to the contained instance, helps when passing to stuff that relies on the original instance signature. We guarantee it's always present due to how keybindings were designed.

Reply:
> It comes from the iterator definition, just a reference to the contained instance, helps when passing to stuff that relies on the original instance signature. We guarantee it's always present due to how keybindings were designed.

Yes, sorry, I meant "what does including this step do that wouldn't be achieved by just dereferencing the iterator directly". As far as I can tell `it` is of type `std::vector<Action>::iterator` and `a` (aka `it.base()`) is of type `Action *`, which for the purposes of accessing the value contained should make no difference, so skipping `a` altogether and using `*it` where currently there's `*a` should have the same result (see the suggestion). Is that not the case?

Reply:
> What would you call it instead?

If it were up to me I'd remove the alias and just use `std::vector <Action>` so that it would be clear what to expect from the container, but conventions etc are up to engine maintainer decision

Reply:
That's reasonable from my part, I don't mind changing it. Alias were only chosen because this seems to be a common convention (at least for the KeyBindings related code sections).